keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,1055 @@
+goroutine to finish, and it will wait until the deadline associated 
+with the request `Context` exceeds. The following gist shows the essence
+of the `timeout` filter. 
+
+```
+	resultCh := make(chan interface{})
+	go func() {
+		defer func() {
+			err := recover()
+			resultCh <- err
+		}()
+		handler.ServeHTTP(w, req)
+	}()
+	
+	ctx := req.Context()
+	select {
+	case  <-resultCh:
+		return
+	case <-ctx.Done():
+		// the request has timed out, return a 504 Status Code to the client
+	}
+```
+
+
+### Eanbling per-request Read/Write Deadline
+
+Primarily, there are three steps:
+
+- a) Enablement must be backed by the feature gate `PerHandlerReadWriteTimeout`
+  so that it can be toggled in its entirety.
+- b) Enable per-request read/write deadline for each non long-running request.
+- c) Remove the `timeout` filter from the handler chain, since `b` makes it redundant.
+
+> Setting PerHandlerReadWriteTimeout=false must revert both `b` and `c`
+
+
+We can utilize the deadline associated with the `Context` to set both the read and
+write timeout for a given non long-running request.
+
+Inside the `WithRequestDeadline` filter, we can enable per-request timeout, as shown below:
+```
+if utilfeature.DefaultFeatureGate.Enabled(features.PerHandlerReadWriteTimeout) {
+	ctrl := http.NewResponseController(w)
+	if err := ctrl.SetReadDeadline(deadline); err != nil {
+		handleError(w, req, http.StatusInternalServerError, "failed to set read deadline", err)
+			return
+		}
+	if err := ctrl.SetWriteDeadline(deadline); err != nil {
+		handleError(w, req, http.StatusInternalServerError, "failed to set write deadline", err)
+		return
+	}
+}
+```
+
+Alternatively, we can encapsulate the per-request read/write timeout logic into
+its own filter that is wrapped by `WithRequestDeadline`.
+```
+   handler = WithPerRequestTimeout(handler, ...)
+   handler = WithRequestDeadline(handler, ...)
+```
+For the remainder of this proposal, we are going to refer to `WithPerRequestTimeout`.
+
+
+> There is no known use case today where there is a need to manipulate the 
+> read/write deadline for anya given request, so for the initial implementation, we
+> are setting the read and write deadline once inside the `WithPerRequestTimeout`
+> filter, and not providing any knob for it to be manipulated further down.
+
+This is the [issue](https://github.com/golang/go/issues/54136) that tracks the implementation of `ResponseController`
+
+Let's delve into the inner working of read/write deadline for `http2`:
+
+##### Write Deadline
+
+`SetWriteDeadline` sets up a `time.AfterFunc` [1] with the given 
+deadline, once the write deadline exceeds the stream is closed [2] with a `stream reset`
+error, and the client sees the stream error immediately, although the request 
+handler on the server might still be running after the timeout happens.
+
+If the request handler on the server continues to write to the underlying
+`ResponseWriter` object, the `Write` method will eventually return an
```

Comment:
so it's only eventually if the buffer is filled?

Reply:
yes, although the error from `FlushError` is immediate even if the buffer is not full
```
w.Write([]byte("a"))
w.FlushError()
```